File: clients/ember/tests/integration/e2e.test.js
/**
 * End-to-end tests for Ember SDK
 *
 * Uses the shared test-site (FluffyCloud) for consistent testing across all SDKs.
 * These tests verify the full flow:
 * 1. Start TDD server
 * 2. Start test-site server
 * 3. Launch browser via our Playwright-based launcher
 * 4. Capture screenshots and verify they reach the TDD server
 *
 * Run with: RUN_E2E=1 npm run test:integration
 */

import assert from 'node:assert';
import { spawn } from 'node:child_process';
import { existsSync, mkdirSync, rmSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join, resolve } from 'node:path';
import { after, before, describe, it } from 'node:test';
import { closeBrowser, launchBrowser } from '../../src/launcher/browser.js';
import {
  setPage,
  startScreenshotServer,
  stopScreenshotServer,
} from '../../src/launcher/screenshot-server.js';

// Paths
let testDir = join(tmpdir(), `vizzly-ember-test-${Date.now()}`);
let testSitePath = resolve(import.meta.dirname, '../../../../test-site');

// Helper to start a static file server for test-site
function startTestSiteServer(port = 3030) {
  return new Promise((resolve, reject) => {
    let server = spawn('python3', ['-m', 'http.server', String(port)], {
      cwd: testSitePath,
      stdio: ['ignore', 'pipe', 'pipe'],
    });

    let resolved = false;

    server.stderr.on('data', data => {
      let msg = data.toString();
      if (msg.includes('Serving HTTP') && !resolved) {
        resolved = true;
        resolve({ server, port, url: `http://127.0.0.1:${port}` });
      }
    });

    server.on('error', err => {
      if (!resolved) {
        resolved = true;
        reject(err);
      }
    });

    // Fallback timeout
    setTimeout(() => {
      if (!resolved) {
        resolved = true;
        resolve({ server, port, url: `http://127.0.0.1:${port}` });
      }
    }, 2000);
  });
}

// Check if running under `vizzly tdd run` or `vizzly run`
let externalServer = !!process.env.VIZZLY_SERVER_URL;

// =============================================================================
// E2E Tests with TDD Server and Test Site
// =============================================================================

describe('e2e with TDD server using shared test-site', { skip: !process.env.RUN_E2E }, () => {
  let tddServer = null;
  let testSiteServer = null;
  let screenshotServer = null;
  let browserInstance = null;
  let testSiteUrl = null;

  before(async () => {
    // Check test-site exists
    assert.ok(
      existsSync(join(testSitePath, 'index.html')),
      'test-site/index.html should exist'
    );

    // Start test-site server
    let testSiteInfo = await startTestSiteServer(3030 + Math.floor(Math.random() * 1000));
    testSiteServer = testSiteInfo.server;
    testSiteUrl = testSiteInfo.url;

    // Start TDD server only if not running under vizzly wrapper
    if (!externalServer) {
      // Create test directory for TDD server
      mkdirSync(testDir, { recursive: true });

      tddServer = spawn('npx', ['vizzly', 'tdd', 'start'], {
        cwd: testDir,
        stdio: ['ignore', 'pipe', 'pipe'],
        env: { ...process.env, VIZZLY_HOME: testDir },
      });

      // Wait for TDD server to start
      await new Promise((resolve, reject) => {
        let timeout = setTimeout(() => reject(new Error('TDD server timeout')), 10000);

        tddServer.stdout.on('data', data => {
          if (
            data.toString().includes('TDD server started') ||
            data.toString().includes('localhost:47392')
          ) {
            clearTimeout(timeout);
            resolve();
          }
        });

        tddServer.on('error', err => {
          clearTimeout(timeout);
          reject(err);
        });
      });
    }

    // Start screenshot server
    screenshotServer = await startScreenshotServer();
  });

  after(async () => {
    if (browserInstance) await closeBrowser(browserInstance);
    if (screenshotServer) await stopScreenshotServer(screenshotServer);
    if (testSiteServer) testSiteServer.kill('SIGTERM');
    if (tddServer && !externalServer) {
      tddServer.kill('SIGTERM');
      await new Promise(resolve => {
        tddServer.on('exit', resolve);
        setTimeout(resolve, 2000);
      });
    }
    if (!externalServer) {
      try {
        rmSync(testDir, { recursive: true, force: true });
      } catch {
        // Ignore cleanup errors
      }
    }
  });

  // ===========================================================================
  // Homepage Tests
  // ===========================================================================

  it('captures homepage full page screenshot', async () => {
    let screenshotUrl = `http://127.0.0.1:${screenshotServer.port}`;

    browserInstance = await launchBrowser('chromium', `${testSiteUrl}/index.html`, {
      screenshotUrl,
      playwrightOptions: { headless: true },
    });

    setPage(browserInstance.page);

    let response = await fetch(`${screenshotUrl}/screenshot`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        name: 'homepage-full',
        properties: { page: 'homepage', fullPage: true },
      }),
    });

    let result = await response.json();
    assert.strictEqual(response.status, 200, 'Should succeed');
    assert.ok(['new', 'match'].includes(result.status), `Should have status 'new' or 'match', got: ${result.status}`);
  });

  it('captures navigation bar with selector', async () => {
    let screenshotUrl = `http://127.0.0.1:${screenshotServer.port}`;

    let response = await fetch(`${screenshotUrl}/screenshot`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        name: 'homepage-nav',
        selector: 'nav',
        properties: { component: 'navigation' },
      }),
    });

    let result = await response.json();
    assert.strictEqual(response.status, 200, 'Should succeed');
    assert.ok(['new', 'match'].includes(result.status), `Should have status 'new' or 'match', got: ${result.status}`);
  });

  it('captures hero section', async () => {
    let screenshotUrl = `http://127.0.0.1:${screenshotServer.port}`;

    let response = await fetch(`${screenshotUrl}/screenshot`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        name: 'homepage-hero',
        selector: 'section',
        properties: { section: 'hero' },
      }),
    });

    let result = await response.json();
    assert.strictEqual(response.status, 200, 'Should succeed');
    assert.ok(['new', 'match'].includes(result.status), `Should have status 'new' or 'match', got: ${result.status}`);
  });

  // ===========================================================================
  // Multiple Pages
  // ===========================================================================

  it('captures features page', async () => {
    let screenshotUrl = `http://127.0.0.1:${screenshotServer.port}`;

    // Navigate to features page
    await browserInstance.page.goto(`${testSiteUrl}/features.html`);
    await browserInstance.page.waitForLoadState('networkidle');

    let response = await fetch(`${screenshotUrl}/screenshot`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        name: 'features-full',
        properties: { page: 'features' },
      }),
    });

    let result = await response.json();
    assert.strictEqual(response.status, 200, 'Should succeed');
    assert.ok(['new', 'match'].includes(result.status), `Should have status 'new' or 'match', got: ${result.status}`);
  });

  it('captures pricing page', async () => {
    let screenshotUrl = `http://127.0.0.1:${screenshotServer.port}`;

    await browserInstance.page.goto(`${testSiteUrl}/pricing.html`);
    await browserInstance.page.waitForLoadState('networkidle');

    let response = await fetch(`${screenshotUrl}/screenshot`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        name: 'pricing-full',
        properties: { page: 'pricing' },
      }),
    });

    let result = await response.json();
    assert.strictEqual(response.status, 200, 'Should succeed');
    assert.ok(['new', 'match'].includes(result.status), `Should have status 'new' or 'match', got: ${result.status}`);
  });

  it('captures contact page', async () => {
    let screenshotUrl = `http://127.0.0.1:${screenshotServer.port}`;

    await browserInstance.page.goto(`${testSiteUrl}/contact.html`);
    await browserInstance.page.waitForLoadState('networkidle');

    let response = await fetch(`${screenshotUrl}/screenshot`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        name: 'contact-full',
        properties: { page: 'contact' },
      }),
    });

    let result = await response.json();
    assert.strictEqual(response.status, 200, 'Should succeed');
    assert.ok(['new', 'match'].includes(result.status), `Should have status 'new' or 'match', got: ${result.status}`);
  });

  // ===========================================================================
  // Screenshot Options
  // ===========================================================================

  it('captures screenshot with threshold option', async () => {
    let screenshotUrl = `http://127.0.0.1:${screenshotServer.port}`;

    await browserInstance.page.goto(`${testSiteUrl}/index.html`);

    let response = await fetch(`${screenshotUrl}/screenshot`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        name: 'threshold-test',
        selector: 'nav',
        threshold: 5,
        properties: { test: 'threshold' },
      }),
    });

    let result = await response.json();
    assert.strictEqual(response.status, 200, 'Should succeed');
    assert.ok(['new', 'match'].includes(result.status), `Should have status 'new' or 'match', got: ${result.status}`);
  });

  it('captures screenshot with all options', async () => {
    let screenshotUrl = `http://127.0.0.1:${screenshotServer.port}`;

    let response = await fetch(`${screenshotUrl}/screenshot`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        name: 'all-options-test',
        selector: 'footer',
        threshold: 3,
        properties: {
          browser: 'chromium',
          viewport: { width: 1920, height: 1080 },
          testType: 'comprehensive',
        },
      }),
    });

    let result = await response.json();
    assert.strictEqual(response.status, 200, 'Should succeed');
    assert.ok(['new', 'match'].includes(result.status), `Should have status 'new' or 'match', got: ${result.status}`);
  });

  // ===========================================================================
  // Multiple Screenshots Per Test
  // ===========================================================================

  it('captures multiple screenshots in sequence', async () => {
    let screenshotUrl = `http://127.0.0.1:${screenshotServer.port}`;
    let pages = ['index.html', 'features.html', 'pricing.html'];

    for (let i = 0; i < pages.length; i++) {
      await browserInstance.page.goto(`${testSiteUrl}/${pages[i]}`);
      await browserInstance.page.waitForLoadState('networkidle');

      let response = await fetch(`${screenshotUrl}/screenshot`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          name: `sequence-${i}`,
          selector: 'nav',
          properties: { page: pages[i], index: i },
        }),
      });

      let result = await response.json();
      assert.strictEqual(response.status, 200, `Should succeed for ${pages[i]}`);
      assert.ok(['new', 'match'].includes(result.status), `Should have status 'new' or 'match', got: ${result.status}`);
    }
  });
});

// =============================================================================
// Tests without TDD Server (verify screenshot server mechanics)
// Skip this suite when running under vizzly wrapper since a TDD server IS available
// =============================================================================

describe('screenshot server mechanics (without TDD server)', { skip: externalServer }, () => {
  let screenshotServer = null;
  let testSiteServer = null;
  let browserInstance = null;
  let testSiteUrl = null;

  before(async () => {
    // Start test-site server
    let testSiteInfo = await startTestSiteServer(4030 + Math.floor(Math.random() * 1000));
    testSiteServer = testSiteInfo.server;
    testSiteUrl = testSiteInfo.url;

    // Start screenshot server
    screenshotServer = await startScreenshotServer();
    let screenshotUrl = `http://127.0.0.1:${screenshotServer.port}`;

    // Launch browser pointing to test-site
    browserInstance = await launchBrowser('chromium', `${testSiteUrl}/index.html`, {
      screenshotUrl,
      playwrightOptions: { headless: true },
    });

    setPage(browserInstance.page);
  });

  after(async () => {
    if (browserInstance) await closeBrowser(browserInstance);
    if (screenshotServer) await stopScreenshotServer(screenshotServer);
    if (testSiteServer) testSiteServer.kill('SIGTERM');
  });

  it('screenshot server receives request and captures screenshot (fails without TDD)', async () => {
    let response = await fetch(`http://127.0.0.1:${screenshotServer.port}/screenshot`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        name: 'no-tdd-test',
        selector: 'nav',
      }),
    });

    // Without TDD server, should fail at forwarding step
    assert.strictEqual(response.status, 500, 'Should fail without TDD server');

    let result = await response.json();
    assert.ok(
      result.error.includes('No Vizzly server found'),
      'Error should mention missing server'
    );
  });

  it('health endpoint works while page is set', async () => {
    let response = await fetch(`http://127.0.0.1:${screenshotServer.port}/health`);
    let result = await response.json();

    assert.strictEqual(response.status, 200);
    assert.strictEqual(result.status, 'ok');
    assert.strictEqual(result.page, true, 'Page should be available');
  });
});

File: clients/ruby/test/e2e_test.rb
# frozen_string_literal: true

require 'minitest/autorun'
require 'json'
require 'fileutils'
require 'tmpdir'
require 'webrick'
require_relative '../lib/vizzly'

# Helper methods for E2E test setup and teardown
module E2ETestHelpers
  def find_vizzly_cli
    path = File.expand_path('../../../dist/cli.js', __dir__)
    return nil unless File.exist?(path)

    path
  end

  def cli_path
    @cli_path ||= find_vizzly_cli
  end

  def find_test_site
    test_site_path = File.expand_path('../../../test-site', __dir__)
    return nil unless File.exist?(File.join(test_site_path, 'index.html'))

    test_site_path
  end

  def start_test_site_server
    @test_site_port = rand(3030..4029)
    @test_site_url = "http://localhost:#{@test_site_port}"

    @test_site_server = WEBrick::HTTPServer.new(
      Port: @test_site_port,
      DocumentRoot: @test_site_path,
      Logger: WEBrick::Log.new(File::NULL),
      AccessLog: []
    )

    @test_site_thread = Thread.new { @test_site_server.start }
    sleep 0.5
  end

  def stop_test_site_server
    @test_site_server&.shutdown
    @test_site_thread&.join(2)
  end

  def start_vizzly_server
    return if @external_server

    pid = spawn('node', cli_path, 'tdd', 'start', %i[out err] => File::NULL)
    _pid, status = Process.wait2(pid)
    raise 'Failed to start Vizzly TDD server' unless status.success?

    30.times do
      break if File.exist?('.vizzly/server.json')

      sleep 0.1
    end

    raise 'Vizzly server failed to start' unless File.exist?('.vizzly/server.json')
  end

  def stop_vizzly_server
    return if @external_server

    pid = spawn('node', cli_path, 'tdd', 'stop', %i[out err] => File::NULL)
    Process.wait(pid)
  end

  def setup_selenium
    options = Selenium::WebDriver::Chrome::Options.new
    options.add_argument('--headless')
    options.add_argument('--disable-gpu')
    options.add_argument('--window-size=1920,1080')
    options.add_argument('--no-sandbox')
    options.add_argument('--disable-dev-shm-usage')

    @driver = Selenium::WebDriver.for :chrome, options: options
  end

  def wait_for_page_load
    wait = Selenium::WebDriver::Wait.new(timeout: 10)
    wait.until { @driver.find_element(tag_name: 'body') }
    sleep 0.3
  end

  def capture_screenshot(name, options = {})
    image_data = @driver.screenshot_as(:png)
    Vizzly.screenshot(name, image_data, options)
  end

  def capture_element_screenshot(name, element, options = {})
    @driver.execute_script('arguments[0].scrollIntoView(true);', element)
    sleep 0.1

    image_data = element.screenshot_as(:png)
    Vizzly.screenshot(name, image_data, options)
  end

  def assert_screenshot_result(result)
    assert result, 'Expected screenshot result to be non-nil'
    assert %w[new match].include?(result['status']),
           "Expected status 'new' or 'match', got: #{result['status']}"
  end
end

# E2E test using the shared test-site (FluffyCloud)
# Run with: VIZZLY_E2E=1 ruby test/e2e_test.rb
#
# When run via `vizzly tdd run`, VIZZLY_SERVER_URL is set and we use that server.
# When run standalone, we start our own server.
#
# Requires:
# - Selenium WebDriver gem: gem install selenium-webdriver
# - Chrome/Chromium browser
# - ChromeDriver in PATH
#
# This test captures real browser screenshots from the shared test-site
# to ensure consistency with other SDK E2E tests.
class E2ETest < Minitest::Test
  include E2ETestHelpers

  def setup
    skip 'Set VIZZLY_E2E=1 to run E2E tests' unless ENV['VIZZLY_E2E']

    begin
      require 'selenium-webdriver'
    rescue LoadError
      skip 'selenium-webdriver gem not installed (gem install selenium-webdriver)'
    end

    @original_dir = Dir.pwd
    Vizzly.reset!

    # Check if we're running under `vizzly tdd run` or `vizzly run`
    @external_server = !ENV['VIZZLY_SERVER_URL'].nil?

    if @external_server
      # Running under vizzly wrapper - server is already running
      @temp_dir = nil
    else
      # Running standalone - create temp dir and start our own server
      @temp_dir = Dir.mktmpdir
      Dir.chdir(@temp_dir)
      skip 'Vizzly CLI not found' unless cli_path
    end

    @test_site_path = find_test_site
    skip 'test-site not found' unless @test_site_path

    # Start test site server
    start_test_site_server

    # Start Vizzly TDD server (only if not using external)
    start_vizzly_server

    # Setup Selenium
    setup_selenium
  end

  def teardown
    @driver&.quit
    stop_vizzly_server
    stop_test_site_server
    return unless @temp_dir

    Dir.chdir(@original_dir) if @original_dir
    FileUtils.rm_rf(@temp_dir)
  end

  # ===========================================================================
  # Homepage Tests
  # ===========================================================================

  def test_homepage_full_page
    @driver.navigate.to "#{@test_site_url}/index.html"
    wait_for_page_load

    result = capture_screenshot('homepage-full', full_page: true)
    assert_screenshot_result(result)
  end

  def test_homepage_navigation
    @driver.navigate.to "#{@test_site_url}/index.html"
    wait_for_page_load

    nav = @driver.find_element(tag_name: 'nav')
    result = capture_element_screenshot('homepage-nav', nav)
    assert_screenshot_result(result)
  end

  def test_homepage_hero_section
    @driver.navigate.to "#{@test_site_url}/index.html"
    wait_for_page_load

    hero = @driver.find_element(tag_name: 'section')
    result = capture_element_screenshot('homepage-hero', hero,
                                        properties: { section: 'hero', page: 'homepage' })
    assert_screenshot_result(result)
  end

  # ===========================================================================
  # Multiple Pages
  # ===========================================================================

  def test_features_page
    @driver.navigate.to "#{@test_site_url}/features.html"
    wait_for_page_load

    result = capture_screenshot('features-full', full_page: true)
    assert_screenshot_result(result)
  end

  def test_pricing_page
    @driver.navigate.to "#{@test_site_url}/pricing.html"
    wait_for_page_load

    result = capture_screenshot('pricing-full', full_page: true)
    assert_screenshot_result(result)
  end

  def test_contact_page
    @driver.navigate.to "#{@test_site_url}/contact.html"
    wait_for_page_load

    result = capture_screenshot('contact-full', full_page: true)
    assert_screenshot_result(result)
  end

  # ===========================================================================
  # Options Testing
  # ===========================================================================

  def test_screenshot_with_threshold
    @driver.navigate.to "#{@test_site_url}/index.html"
    wait_for_page_load

    nav = @driver.find_element(tag_name: 'nav')
    result = capture_element_screenshot('threshold-test', nav, threshold: 5)
    assert_screenshot_result(result)
  end

  def test_screenshot_with_properties
    @driver.navigate.to "#{@test_site_url}/index.html"
    wait_for_page_load

    result = capture_screenshot('props-test',
                                properties: {
                                  browser: 'chrome',
                                  viewport: { width: 1920, height: 1080 },
                                  theme: 'light'
                                })
    assert_screenshot_result(result)
  end

  def test_screenshot_with_all_options
    @driver.navigate.to "#{@test_site_url}/index.html"
    wait_for_page_load

    result = capture_screenshot('all-options-test',
                                full_page: true,
                                threshold: 3,
                                properties: {
                                  browser: 'chrome',
                                  page: 'homepage',
                                  test_type: 'comprehensive'
                                })
    assert_screenshot_result(result)
  end

  # ===========================================================================
  # Multiple Screenshots
  # ===========================================================================

  def test_navigation_across_pages
    pages = %w[index.html features.html pricing.html contact.html]

    pages.each_with_index do |page, index|
      @driver.navigate.to "#{@test_site_url}/#{page}"
      wait_for_page_load

      nav = @driver.find_element(tag_name: 'nav')
      result = capture_element_screenshot("nav-page-#{index}", nav,
                                          properties: { page: page })
      assert_screenshot_result(result)
    end
  end
end
